slides/D-trait-objects-patterns.md

```diff
@@ -0,0 +1,824 @@
+    };
+
+    log("Hello, world!🦀", &mut logger);
+}
+```
+<v-click>
+
+- *💸 Cost: pointer indirection via vtable &rarr; less performant*
+- *💰 Benefit: no monomorphization &rarr; smaller binary & shorter compile time!*
+</v-click>
+
+::right::
+<!-- TODO switch out this JPEG for an SVG that works both in dark and light theme -->
+<img src="/images/D-trait-object-layout.jpg" style="margin-left:5%; margin-top: 50px; max-width: 100%; max-height: 90%;">
+
+
+---
+layout: default
+---
+
+# Fixing dynamic logger
+
+- Trait objects `&dyn T`, `Box<dyn T>`, ... implement `T`!
+
+```rust{all|9-12|1-2}
+// We no longer require L be `Sized`, so to accept trait objects
+fn log<L: Write + ?Sized>(entry: &str, logger: &mut L) {
+    write!(logger, "{}", entry);
+}
+
+fn main() {
+    let log_file: Option<PathBuf> = 
+        todo!("read args");
+    // Create a trait object that implements `Write`
+    let logger: &mut dyn Write = match log_file {
+        Some(log_path) => &mut FileLogger { log_path },
+        Nome => &mut StdOutLogger,
+    };
+
+    log("Hello, world!🦀", logger);
+}
+```
+And all is well!
+
+---
+layout: default
+---
+
+# Forcing dynamic dispatch
+
+Sometimes you want to enforce API users (or colleagues) to use dynamic dispatch
+
+```rust{all|1}
+fn log(entry: &str, logger: &mut dyn Write) {
+    write!(logger, "{}", entry);
+}
+
+fn main() {
+    let log_file: Option<PathBuf> = 
+        todo!("read args");
+    // Create a trait object that implements `Write`
+    let logger: &mut dyn Write = match log_file {
+        Some(log_path) => &mut FileLogger { log_path },
+        Nome => &mut StdOutLogger,
+    };
+
+
+    log("Hello, world!🦀", &mut logger);
+}
+```
+
+---
+layout: default
+---
+
+# Fixing the renderer
+
+```rust
+fn main() {
+    let mut shapes = Vec::new();
+    let circle = Circle;
+    shapes.push(circle);
+    let rect = Rectangle;
+    shapes.push(rect);
+    shapes.iter().for_each(|shape| shape.paint());
+}
+```
+<v-click>
+Becomes
+
+```rust{all|2,3,5}
+fn main() {
+    let mut shapes: Vec<Box<dyn Render>> = Vec::new();
+    let circle = Box::new(Circle);
+    shapes.push(circle);
+    let rect = Box::new(Rectangle);
+    shapes.push(rect);
+    shapes.iter().for_each(|shape| shape.paint());
+}
+```
+
+All set!
+</v-click>
+
+---
+layout: default
+---
+
+# Trait object limitations
+
+- Pointer indirection cost
+- Harder to debug
+- Type erasure
+- Not all traits work:
+
+*Traits need to be 'Object Safe'*
+
+
+---
+layout: default
+---
+
+# Object safety
+
+In order for a trait to be object safe, these conditions need to be met:
+
+- If `trait T: Y`, then`Y` must be object safe
+- trait `T` must not be `Sized`
+- No associated constants allowed
+- No associated types with generic allowed
+- All associated functions must either be dispatchable from a trait object, or explicitly non-dispatchable
+
+Details in [The Rust Reference](https://doc.rust-lang.org/reference/items/traits.html#object-safety). Read them!
+
```

Comment:
A simple question: *"WHY?"* Now you probably don't want to talk about that (it would probably be 45 minutes lecture in its own right), but the link provided *also* doesn't tell me the "why". Is there a RFC or a book that students can read?

Reply:
I'll add some content on the 'why'